alf/algorithms/particle_vi_algorithm.py

```diff
@@ -0,0 +1,298 @@
+                shape [num_particles] a loss term for optimizing the generator.
+            transform_func (Callable): tranform functoin on particles. Used in 
+                function value based par_vi, where each particle represents 
+                parameters of a neural network function. It is call by
+                transform_func(particles) which returns the following,
+                - outputs: outputs of network parameterized by particles evaluated
+                    on predifined training batch.
+                - extra_outputs: outputs of network parameterized by particles
+                    evaluated on additional sampled data.
+            entropy_regularization (float): weight of the repulsive term in par_vi. 
+                If None, use self._entropy_regularization.
+            loss_mask (Tensor): mask indicating which samples are valid for loss
+                propagation.
+            state: not used
+
+        Returns:
+            AlgorithmStep:
+                outputs: Tensor with shape (num_particles, dim)
+                info: LossInfo
+        """
+        if entropy_regularization is None:
+            entropy_regularization = self._entropy_regularization
+        loss, loss_propagated = self._grad_func(
+            self.particles, loss_func, entropy_regularization, transform_func)
+        if loss_mask is not None:
+            loss_propagated = loss_propagated * loss_mask
+
+        return AlgStep(
+            output=self.particles,
+            state=(),
+            info=LossInfo(loss=loss_propagated, extra=loss))
+
+    def _kernel_width(self, dist):
+        """Update kernel_width averager and get latest kernel_width. """
+        if dist.ndim > 1:
+            dist = torch.sum(dist, dim=-1)
+            assert dist.ndim == 1, "dist must have dimension 1 or 2."
+        width, _ = torch.median(dist, dim=0)
+        width = width / np.log(len(dist))
+        self._kernel_width_averager.update(width)
+
+        return self._kernel_width_averager.get()
+
+    def _rbf_func(self, x, y):
+        r"""
+        Compute the rbf kernel and its gradient w.r.t. first entry 
+        :math:`K(x, y), \nabla_x K(x, y)`, used by svgd_grad.
+
+        Args:
+            x (Tensor): set of N particles, shape (Nx x W), where W is the 
+                dimenseion of each particle
+            y (Tensor): set of N particles, shape (Ny x W), where W is the 
+                dimenseion of each particle
+
+        Returns:
+            :math:`K(x, y)` (Tensor): the RBF kernel of shape (Nx x Ny)
+            :math:`\nabla_x K(x, y)` (Tensor): the derivative of RBF kernel of shape (Nx x Ny x D)
+            
+        """
+        Nx, Dx = x.shape
+        Ny, Dy = y.shape
+        assert Dx == Dy
+        diff = x.unsqueeze(1) - y.unsqueeze(0)  # [Nx, Ny, W]
+        dist_sq = torch.sum(diff**2, -1)  # [Nx, Ny]
+        # h = self._kernel_width(dist_sq.view(-1))
```

Comment:
Since this is not used, perhaps you can remove the related code.

Reply:
Removed.